mobile_verifier/src/coverage.rs

```diff
@@ -0,0 +1,719 @@
+        .await?
+    {
+        return Ok(CoverageObjectValidity::InvalidPubKey);
+    }
+
+    Ok(CoverageObjectValidity::Valid)
+}
+
+#[derive(Clone, FromRow)]
+pub struct HexCoverage {
+    pub uuid: Uuid,
+    pub hex: i64,
+    pub indoor: bool,
+    pub cbsd_id: String,
+    pub signal_level: SignalLevel,
+    pub coverage_claim_time: DateTime<Utc>,
+}
+
+#[derive(Eq)]
+struct CoverageLevel {
+    cbsd_id: String,
+    coverage_claim_time: DateTime<Utc>,
+    hotspot: PublicKeyBinary,
+    indoor: bool,
+    signal_level: SignalLevel,
+}
+
+impl PartialEq for CoverageLevel {
+    fn eq(&self, other: &Self) -> bool {
+        self.coverage_claim_time == other.coverage_claim_time
+    }
+}
+
+impl PartialOrd for CoverageLevel {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.coverage_claim_time.cmp(&other.coverage_claim_time))
+    }
+}
+
+impl Ord for CoverageLevel {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.coverage_claim_time.cmp(&other.coverage_claim_time)
+    }
+}
+
+impl CoverageLevel {
+    fn coverage_points(&self) -> anyhow::Result<Decimal> {
+        Ok(match (self.indoor, self.signal_level) {
+            (true, SignalLevel::High) => dec!(400),
+            (true, SignalLevel::Low) => dec!(100),
+            (false, SignalLevel::High) => dec!(16),
+            (false, SignalLevel::Medium) => dec!(8),
+            (false, SignalLevel::Low) => dec!(4),
+            (_, SignalLevel::None) => dec!(0),
+            _ => anyhow::bail!("Indoor radio cannot have a signal level of medium"),
+        })
+    }
+}
+
+#[derive(PartialEq, Debug)]
+pub struct CoverageReward {
+    pub cbsd_id: String,
+    pub points: Decimal,
+    pub hotspot: PublicKeyBinary,
+}
+
+pub const MAX_RADIOS_PER_HEX: usize = 5;
+
+#[async_trait::async_trait]
+pub trait CoveredHexStream {
+    async fn covered_hex_stream<'a>(
+        &'a self,
+        cbsd_id: &'a str,
+        coverage_obj: &'a Uuid,
+        period_end: DateTime<Utc>,
+    ) -> Result<BoxStream<'a, Result<HexCoverage, sqlx::Error>>, sqlx::Error>;
+}
+
+#[derive(sqlx::FromRow)]
+pub struct Seniority {
+    pub uuid: Uuid,
+    pub seniority_ts: DateTime<Utc>,
+    pub last_heartbeat: DateTime<Utc>,
+    pub inserted_at: DateTime<Utc>,
+}
+
+#[async_trait::async_trait]
+impl CoveredHexStream for Pool<Postgres> {
+    async fn covered_hex_stream<'a>(
+        &'a self,
+        cbsd_id: &'a str,
+        coverage_obj: &'a Uuid,
+        period_end: DateTime<Utc>,
+    ) -> Result<BoxStream<'a, Result<HexCoverage, sqlx::Error>>, sqlx::Error> {
+        // Adjust the coverage
+        let seniority: Seniority = sqlx::query_as(
+            r#"
+            SELECT * FROM seniority
+            WHERE
+              cbsd_id = $1 AND
+              inserted_at <= $2
+            ORDER BY inserted_at DESC
+            LIMIT 1
+            "#,
+        )
+        .bind(cbsd_id)
+        .bind(period_end)
+        .fetch_one(self)
```

Comment:
is always guaranteed to be one? what happens when we first activate the feature and it's populating the new record types?